packages/core/src/common/array-utils.ts

```diff
@@ -106,4 +106,33 @@ export namespace ArrayUtils {
     export function coalesce<T>(array: ReadonlyArray<T | undefined | null>): T[] {
         return <T[]>array.filter(e => !!e);
     }
+
+    /**
+     * A safe variant to push additional items to an array. By default, the
+     * array push operation in combination with the spread operator is used.
+     * However, if the callstack size is exceeded on large additions we use the
+     * concatenation of arrays instead as it not depend on the callstack
+     * size.
+     *
+     * The original array might be modified.
+     *
+     * @param array An array of elements.
+     * @param items Additional elements to be added to the array.
+     * @returns An array containing the original elements with the additional
+     *   elements appended. This may or may not be the array that was handed in.
+     */
+    export function pushAll<T>(array: T[], items: T[]): T[] {
+        try {
+            // typically faster but might fail depending on the number of items and the callstack size
+            array.push(...items);
+            return array;
+        } catch (error) {
```

Comment:
What do you mean exactly by `our version` vs `recursive version`? In our scenario we had a table will millions of data sets and everything was still working very quickly.

Reply:
The simplest depth first tree traversal is recursive. We're using a stack here, for some reason. Since I couldn't find any info _why_ we're using a more complicated version of tree traversal here, I feel it may be a case of premature optimization.

Reply:
I just tested a little bit locally (nothing scientific for sure) against this recursive version:
```
    export function* depthFirst<T>(root: T, ...): IterableIterator<T> {
        yield root;
        const sortedChildren = (children(root) || []).filter(include).reverse();
        for (const child of sortedChildren) {
            yield* depthFirst(child, children);
        }
    }
```

and I'd say with 5.000.000 I saw about 10-15% improvement with the stack-version vs the recursive version. So I think it is useful to keep it like that for now. It also matches the breadth-first search so it is not that hard to read.